esp/services/ws_smc/ws_smcService.hpp

```diff
@@ -154,15 +157,67 @@ class CActivityInfo : public CInterface, implements IInterface
     inline IArrayOf<IEspActiveWorkunit>& queryActiveWUs() { return aws; };
     inline IArrayOf<IEspServerJobQueue>& queryServerJobQueues() { return serverJobQueues; };
     inline IArrayOf<IEspDFUJob>& queryDFURecoveryJobs() { return DFURecoveryJobs; };
+    inline StringBuffer& queryTimeCached(StringBuffer& str) { return timeCached.getString(str, true); }
+};
+
+class CWsSMCEx;
+
+class CActivityInfoReader : public CInterface, implements IThreaded
+{
+    bool stopping = false;
+    bool detached = false;
+    unsigned autoRebuildSeconds = DEFAULTACTIVITYINFOCACHEAUTOREBUILDSECOND;
+    unsigned forceRebuildSeconds = DEFAULTACTIVITYINFOCACHEFORCEBUILDSECOND;
+    Owned<CActivityInfo> activityInfoCache;
+    Semaphore sem;
+    CriticalSection crit;
+    CThreaded threaded;
+    std::atomic<bool> waiting = {false};
+public:
+    CActivityInfoReader(unsigned _autoRebuildSeconds, unsigned _forceRebuildSeconds)
+        : autoRebuildSeconds(_autoRebuildSeconds), forceRebuildSeconds(_forceRebuildSeconds), threaded("ActivityInfoReader")
+    {
+        threaded.init(this);
+    };
+
+    ~CActivityInfoReader()
+    {
+        stopping = true;
+        sem.signal();
+        threaded.join();
+    }
+
+    virtual void threadmain() override;
+    CActivityInfo* getActivityInfo()
+    {
+        CriticalBlock b(crit);
+        assert(activityInfoCache);
```

Comment:
in fact, it could still be null by this time, if the thread hasn't got started...
you could hold the crit before the thread starts, and release after 1st build.

Reply:
But also, what happens if detached and this method is called?
I think in 1st commit - it would loop continuously until re-attached.

What is the expected behaviour here when detached? Is there anything to prevent it being called?

Reply:
so conclusion from offline chat, is that it could be detached at any time, including before the service actually starts, i.e. before any activity info is available.

- If detached and no activityInfoCache available, it must return null and callers need to handle that (not assert)

- If not detached it should block until there's an activityInfoCache instance available OR until detached becomes true.
You can could block it, by holding the crit early at CActivityInfoReader time and ensuring released after the 1st creation.

Alternatively, you could check if 'first' and only block if activityInfoCache is unavailable, until it is

Reply:
Got it